torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -1441,8 +1442,184 @@ void InitXlaModuleBindings(py::module m) {
+            // This will create a new XLATensor with different Id which is not
+            // ideal. However there is not an easy way to get the origional
+            // tensor with the tensor_id.
+            // TODO(JackCaoG): invesgate if we have idle XLATensor.
+            auto tensor = bridge::AtenFromXlaTensor(
+                torch_xla::XLATensor::Create(backend_data));
+            ivalues.emplace_back(tensor);
+          }
+          return std::make_pair(tensor_ids, ivalues);
+        });
+
+  // Return true if value of the tensor requires a computation.
+  m.def("_check_tensor_need_materialization",
+        [](const std::vector<at::Tensor>& tensors) -> std::vector<bool> {
+          std::vector<bool> need_materialization;
+          need_materialization.reserve(tensors.size());
+          for (auto& tensor : tensors) {
+            auto xtensor = bridge::TryGetXlaTensor(tensor);
+            if (!xtensor) {
+              // input tensor is not a xla tensor
+              need_materialization.push_back(false);
+            } else if (xtensor->CurrentXlaData() != nullptr) {
+              // input tensor has xla_data which means it is already on device
+              need_materialization.push_back(true);
+            } else if (xtensor->CurrentIrValue().node != nullptr) {
+              torch::lazy::NodePtr node = xtensor->CurrentIrValue().node;
+              if (torch_xla::DeviceData::Cast(
+                      xtensor->CurrentIrValue().node.get()) != nullptr) {
+                need_materialization.push_back(false);
+              } else {
+                // input tensor is an IR other than DeviceData which means a
+                // compuation is required to get the value of this tensor.
+                need_materialization.push_back(true);
+              }
+            } else {
+              // TODO: maybe also handle it is a XLATensor with tensor_data case
+              XLA_CHECK(false);
+            }
+          }
+          return need_materialization;
+        });
+
+  m.def("_get_graph_hash", [](const std::vector<at::Tensor>& tensors) {
+    std::vector<XLATensorPtr> xtensors;
+    xtensors.reserve(tensors.size());
+    for (auto& tensor : tensors) {
+      xtensors.push_back(bridge::GetXlaTensor(tensor));
+    }
+    auto hash = XLATensor::GetGraphHash(xtensors);
+    std::string bin((const char*)&hash, sizeof(hash));
+    return py::bytes(bin);
+  });
+
+  m.def("_run_cached_graph",
+        [](const std::string& hash_str,
+           const std::vector<at::IValue>& graph_inputs)
+            -> std::vector<at::Tensor> {
+          TORCH_CHECK(hash_str.size() == sizeof(torch::lazy::hash_t));
+          torch::lazy::hash_t hash = *(torch::lazy::hash_t*)(hash_str.c_str());
+          auto cachedComputation = XLATensor::GetComputationCache()->Get(hash);
+          // TODO implement a fallback mechanism, or make sure those entries
+          // never get kicked out
```

Comment:
@shunting314 I think we can have a larger default cache size for dynamo, but a fallback sounds more robust solution. How does non-xla backend handles it? Do they never clear the cache?

Reply:
One way to handle this is to propagate a python exception up to dynamo so it can fall back to eager / or trigger recompile. But it's not there yet.

I'm thinking maybe we can have some statistics for how many entries a normal model need first. If the number is usually quite small, then a fallback here should be something good to have but not on the critical path. What do you think?

Reply:
The ts backend does similar thing. We never clear cache explicitly.

Reply:
for the most part cache = CompileTime. It is usually < 50, for the most working models, it is < 25. However for dynamo I am not entirely sure since the tracing mechanism is slight different but I don't expect this to be a huge number (otherwise compilation will take forever).

Reply:
the default cache size is 1024 according to https://github.com/pytorch/xla/blob/d3b1a7db6098b9181d2251f015f5992218d4c38b/torch_xla/csrc/tensor.cpp#L1366

I think we should be fine now, but I will leave this TODO here.